File: lexflow-web/src/lexflow_web/visualization.py
"""Visualization service for converting workflows to tree structures."""

from typing import Any


def workflow_to_tree(workflow_data: dict) -> dict:
    """Convert workflow dict to tree structure for web rendering."""
    workflows = workflow_data.get("workflows", [])

    if not workflows:
        # Check if it's a single implicit workflow (keys at root)
        if "nodes" in workflow_data:
            return {
                "type": "project",
                "workflows": [_build_workflow_tree(workflow_data)],
            }
        return {"error": "No workflows found"}

    # Build tree for all workflows
    workflow_trees = []
    main_interface = {"inputs": [], "outputs": []}

    for w in workflows:
        tree = _build_workflow_tree(w)
        workflow_trees.append(tree)
        if tree.get("name") == "main" or not main_interface["inputs"]:
            main_interface = tree.get("interface", {})

    return {"type": "project", "workflows": workflow_trees, "interface": main_interface}


def _build_workflow_tree(workflow: dict) -> dict:
    """Build tree for a single workflow."""
    nodes = workflow.get("nodes", {})
    interface = workflow.get("interface", {})

    tree = {
        "type": "workflow",
        "name": workflow.get("name", "main"),
        "interface": {
            "inputs": interface.get("inputs", []),
            "outputs": interface.get("outputs", []),
        },
        "variables": workflow.get("variables", {}),
        "children": [],
        "orphans": [],
    }

    # Track all visited nodes (including branch nodes and reporters)
    all_visited = set()

    # Follow node chain from start
    start_node = nodes.get("start", {})
    current_id = start_node.get("next")

    while current_id and current_id not in all_visited:
        all_visited.add(current_id)
        node = nodes.get(current_id)
        if not node:
            break

        tree_node = _node_to_tree(current_id, node, nodes, all_visited)
        tree["children"].append(tree_node)
        current_id = node.get("next")

    # Find orphan nodes (nodes not in the connected chain)
    # First, collect reporter IDs from ALL potential orphan nodes
    # This ensures that if orphan A references orphan B as a reporter,
    # B won't appear as a separate orphan
    all_node_ids = set(nodes.keys()) - {"start"}
    potential_orphan_ids = all_node_ids - all_visited

    for node_id in potential_orphan_ids:
        node = nodes.get(node_id, {})
        _collect_reporter_ids(node.get("inputs", {}), nodes, all_visited)

    # Now recalculate orphan_ids with reporter references accounted for
    orphan_ids = all_node_ids - all_visited

    # Build orphan chains: find chain heads (orphans not pointed to by other orphans)
    # and follow their next pointers to maintain chain order
    orphan_next_targets = set()
    for node_id in orphan_ids:
        node = nodes.get(node_id, {})
        next_id = node.get("next")
        if next_id and next_id in orphan_ids:
            orphan_next_targets.add(next_id)

    # Chain heads are orphans that no other orphan points to
    orphan_heads = [oid for oid in orphan_ids if oid not in orphan_next_targets]

    # Process each orphan chain starting from its head
    processed_orphans = set()
    for head_id in sorted(orphan_heads):
        current_id = head_id
        while current_id and current_id not in processed_orphans:
            node = nodes.get(current_id)
            if not node:
                break
            # Skip reporter nodes (they're embedded in other nodes)
            if node.get("isReporter"):
                processed_orphans.add(current_id)
                current_id = (
                    node.get("next") if node.get("next") in orphan_ids else None
                )
                continue

            processed_orphans.add(current_id)
            orphan_tree = _node_to_tree(current_id, node, nodes, all_visited)

            # Include next pointer if it points to another orphan (for chain visualization)
            next_id = node.get("next")
            if next_id and next_id in orphan_ids:
                orphan_tree["next"] = next_id

            tree["orphans"].append(orphan_tree)
            current_id = next_id if next_id in orphan_ids else None

    return tree


def _node_to_tree(node_id: str, node: dict, all_nodes: dict, all_visited: set) -> dict:
    """Convert a single node to tree node.

    Args:
        all_visited: Set that accumulates ALL visited node IDs (mutated in place)
    """
    opcode = node.get("opcode", "")
    inputs = node.get("inputs", {})
    is_reporter = node.get("isReporter", False)

    # Track reporter nodes referenced in inputs
    _collect_reporter_ids(inputs, all_nodes, all_visited)

    tree_node = {
        "id": node_id,
        "type": _get_node_type(opcode),
        "opcode": opcode,
        "isReporter": is_reporter,
        "inputs": _format_inputs(inputs, all_nodes),
        "children": [],
    }

    # Handle control flow branches
    if opcode in ("control_for", "control_foreach", "control_while"):
        tree_node["config"] = _extract_loop_config(opcode, inputs, all_nodes)
        body_input = inputs.get("BODY", inputs.get("body", {}))
        body_branch = body_input.get("branch") if isinstance(body_input, dict) else None
        if body_branch:
            tree_node["children"].append(
                _build_branch("BODY", body_branch, all_nodes, all_visited)
            )

    elif opcode in ("control_if", "control_if_else"):
        then_input = inputs.get("THEN", inputs.get("then", {}))
        then_branch = then_input.get("branch") if isinstance(then_input, dict) else None
        if then_branch:
            tree_node["children"].append(
                _build_branch("THEN", then_branch, all_nodes, all_visited)
            )

        else_input = inputs.get("ELSE", inputs.get("else", {}))
        else_branch = else_input.get("branch") if isinstance(else_input, dict) else None
        if else_branch:
            tree_node["children"].append(
                _build_branch("ELSE", else_branch, all_nodes, all_visited)
            )

    elif opcode == "control_fork":
        tree_node["children"] = _extract_fork_branches(inputs, all_nodes, all_visited)

    elif opcode == "control_try":
        tree_node["children"] = _extract_try_branches(inputs, all_nodes, all_visited)

    return tree_node


def _collect_reporter_ids(inputs: dict, all_nodes: dict, all_visited: set) -> None:
    """Collect all reporter node IDs referenced in inputs."""
    for value in inputs.values():
        if isinstance(value, dict):
            if "node" in value:
                node_id = value["node"]
                if node_id not in all_visited:
                    all_visited.add(node_id)
                    node = all_nodes.get(node_id, {})
                    # Recursively collect from reporter's inputs
                    node_inputs = node.get("inputs", {})
                    _collect_reporter_ids(node_inputs, all_nodes, all_visited)
            elif "branch" in value:
                # Branch references are handled separately
                pass


def _get_node_type(opcode: str) -> str:
    """Determine node type from opcode."""
    if opcode.startswith("control_"):
        return "control_flow"
    elif opcode.startswith("data_"):
        return "data"
    elif opcode.startswith("io_"):
        return "io"
    elif opcode.startswith("operator_"):
        return "operator"
    elif opcode.startswith("workflow_"):
        return "workflow_op"
    return "opcode"


def _format_inputs(inputs: dict, all_nodes: dict) -> dict:
    """Format inputs for display, resolving references."""
    formatted = {}
    for key, value in inputs.items():
        # Skip branch inputs for control flow (handled separately)
        if key in ("BODY", "body", "THEN", "then", "ELSE", "else", "TRY", "FINALLY"):
            continue
        if key.startswith("CATCH") or key.startswith("BRANCH"):
            continue
        if key in ("HANDLERS", "handlers", "BRANCHES", "branches"):
            continue

        formatted[key] = _format_value(value, all_nodes)
    return formatted


def _format_value(value: Any, all_nodes: dict, depth: int = 0) -> dict:
    """Format a value for display. No depth limit - show full reporter tree."""

    if isinstance(value, dict):
        if "literal" in value:
            return {"type": "literal", "value": value["literal"]}
        elif "variable" in value:
            return {"type": "variable", "name": value["variable"]}
        elif "node" in value:
            node_id = value["node"]
            node = all_nodes.get(node_id, {})
            opcode = node.get("opcode", "")
            node_inputs = node.get("inputs", {})
            return {
                "type": "reporter",
                "id": node_id,
                "opcode": opcode,
                "inputs": {
                    k: _format_value(v, all_nodes, depth + 1)
                    for k, v in node_inputs.items()
                    if not k.startswith("CATCH") and k not in ("BODY", "THEN", "ELSE")
                },
            }
        elif "branch" in value:
            return {"type": "branch", "target": value["branch"]}
        elif "workflow_call" in value:
            return {"type": "workflow_call", "name": value["workflow_call"]}
        else:
            return {"type": "dict", "value": value}
    else:
        return {"type": "literal", "value": value}


def _extract_loop_config(opcode: str, inputs: dict, all_nodes: dict) -> dict:
    """Extract loop configuration."""
    config = {}

    if opcode == "control_for":
        config["var"] = _get_raw_value(inputs.get("VAR", inputs.get("var", "i")))
        config["start"] = _get_raw_value(inputs.get("START", inputs.get("start", 0)))
        config["end"] = _get_raw_value(inputs.get("END", inputs.get("end", 0)))
        step = inputs.get("STEP", inputs.get("step"))
        if step is not None:
            config["step"] = _get_raw_value(step)

    elif opcode == "control_foreach":
        config["var"] = _get_raw_value(inputs.get("VAR", inputs.get("var", "item")))
        config["iterable"] = _format_value(
            inputs.get("ITERABLE", inputs.get("iterable", [])), all_nodes
        )

    elif opcode == "control_while":
        config["condition"] = _format_value(
            inputs.get("CONDITION", inputs.get("condition", True)), all_nodes
        )

    return config


def _get_raw_value(value: Any) -> Any:
    """Extract raw value from possible wrapper."""
    if isinstance(value, dict) and "literal" in value:
        return value["literal"]
    return value


def _build_branch(name: str, start_id: str, all_nodes: dict, all_visited: set) -> dict:
    """Build a branch subtree.

    Args:
        all_visited: Set that accumulates ALL visited node IDs (mutated in place)
    """
    branch = {
        "type": "branch",
        "name": name,
        "children": [],
    }

    current_id = start_id
    while current_id and current_id not in all_visited:
        all_visited.add(current_id)
        node = all_nodes.get(current_id)
        if not node:
            break
        branch["children"].append(
            _node_to_tree(current_id, node, all_nodes, all_visited)
        )
        current_id = node.get("next")

    return branch


def _extract_fork_branches(inputs: dict, all_nodes: dict, all_visited: set) -> list:
    """Extract branches from fork node."""
    branches = []

    # Try list format first
    branch_list = inputs.get("BRANCHES", inputs.get("branches", []))
    if branch_list:
        for i, branch_ref in enumerate(branch_list, 1):
            branch_id = (
                branch_ref.get("branch") if isinstance(branch_ref, dict) else None
            )
            if branch_id:
                branches.append(
                    _build_branch(f"BRANCH{i}", branch_id, all_nodes, all_visited)
                )
    else:
        # Try individual keys
        i = 1
        while f"BRANCH{i}" in inputs:
            branch_ref = inputs[f"BRANCH{i}"]
            branch_id = (
                branch_ref.get("branch") if isinstance(branch_ref, dict) else None
            )
            if branch_id:
                branches.append(
                    _build_branch(f"BRANCH{i}", branch_id, all_nodes, all_visited)
                )
            i += 1

    return branches


def _extract_try_branches(inputs: dict, all_nodes: dict, all_visited: set) -> list:
    """Extract branches from try-catch-finally node."""
    branches = []

    # TRY body
    try_input = inputs.get("TRY", inputs.get("BODY", inputs.get("body", {})))
    try_branch = try_input.get("branch") if isinstance(try_input, dict) else None
    if try_branch:
        branches.append(_build_branch("TRY", try_branch, all_nodes, all_visited))

    # CATCH handlers
    handlers = inputs.get("HANDLERS", inputs.get("handlers", []))
    if handlers:
        for i, handler in enumerate(handlers, 1):
            if isinstance(handler, dict):
                handler_branch = handler.get("branch")
                if handler_branch:
                    branch = _build_branch(
                        f"CATCH{i}", handler_branch, all_nodes, all_visited
                    )
                    branch["exception_type"] = handler.get(
                        "exception_type", "Exception"
                    )
                    branch["var_name"] = handler.get("var_name")
                    branches.append(branch)
    else:
        # Try individual CATCH keys
        i = 1
        while f"CATCH{i}" in inputs:
            catch_input = inputs[f"CATCH{i}"]
            if isinstance(catch_input, dict):
                body_info = catch_input.get("body", {})
                handler_branch = (
                    body_info.get("branch") if isinstance(body_info, dict) else None
                )
                if handler_branch:
                    branch = _build_branch(
                        f"CATCH{i}", handler_branch, all_nodes, all_visited
                    )
                    branch["exception_type"] = catch_input.get(
                        "exception_type", "Exception"
                    )
                    branch["var_name"] = catch_input.get(
                        "var", catch_input.get("var_name")
                    )
                    branches.append(branch)
            i += 1

    # FINALLY
    finally_input = inputs.get("FINALLY", inputs.get("finally", {}))
    finally_branch = (
        finally_input.get("branch") if isinstance(finally_input, dict) else None
    )
    if finally_branch:
        branches.append(
            _build_branch("FINALLY", finally_branch, all_nodes, all_visited)
        )

    return branches

File: lexflow-web/frontend/src/services/visualization.ts
// Visualization service for converting workflows to tree structures
// Uses grammar-driven extraction for dynamic node rendering

import type {
  WorkflowTree,
  WorkflowNode,
  TreeNode,
  BranchNode,
  FormattedValue,
  LoopConfig,
  NodeType,
  WorkflowInterface,
} from "../api/types";
import { getConstruct, getGrammar } from "./grammar";
import type { Construct } from "../types/grammar";

interface WorkflowData {
  workflows?: WorkflowDefinition[];
  nodes?: Record<string, NodeDefinition>;
  name?: string;
  interface?: { inputs?: string[]; outputs?: string[] };
  variables?: Record<string, unknown>;
}

interface WorkflowDefinition {
  name?: string;
  nodes?: Record<string, NodeDefinition>;
  interface?: { inputs?: string[]; outputs?: string[] };
  variables?: Record<string, unknown>;
}

interface NodeDefinition {
  opcode?: string;
  inputs?: Record<string, unknown>;
  next?: string;
  isReporter?: boolean;
}

interface InputValue {
  literal?: unknown;
  variable?: string;
  node?: string;
  branch?: string;
  workflow_call?: string;
}

interface CatchHandler {
  branch?: string;
  exception_type?: string;
  var_name?: string;
  var?: string;
  body?: InputValue;
}

export function workflowToTree(
  workflowData: WorkflowData,
): WorkflowTree | { error: string } {
  const workflows = workflowData.workflows || [];

  if (workflows.length === 0) {
    // Check if it's a single implicit workflow (keys at root)
    if (workflowData.nodes) {
      return {
        type: "project",
        workflows: [buildWorkflowTree(workflowData as WorkflowDefinition)],
      };
    }
    return { error: "No workflows found" };
  }

  // Build tree for all workflows
  const workflowTrees: WorkflowNode[] = [];
  let mainInterface: WorkflowInterface = { inputs: [], outputs: [] };

  for (const w of workflows) {
    const tree = buildWorkflowTree(w);
    workflowTrees.push(tree);
    if (tree.name === "main" || mainInterface.inputs.length === 0) {
      mainInterface = tree.interface;
    }
  }

  return {
    type: "project",
    workflows: workflowTrees,
    interface: mainInterface,
  };
}

function buildWorkflowTree(workflow: WorkflowDefinition): WorkflowNode {
  const nodes = workflow.nodes || {};
  const iface = workflow.interface || {};

  const tree: WorkflowNode = {
    type: "workflow",
    name: workflow.name || "main",
    interface: {
      inputs: iface.inputs || [],
      outputs: iface.outputs || [],
    },
    variables: workflow.variables || {},
    children: [],
    orphans: [],
  };

  // Track all visited nodes (including branch nodes and reporters)
  const allVisited = new Set<string>();

  // Follow node chain from start
  const startNode = nodes["start"] || {};
  let currentId = startNode.next;

  while (currentId && !allVisited.has(currentId)) {
    allVisited.add(currentId);
    const node = nodes[currentId];
    if (!node) break;

    const treeNode = nodeToTree(currentId, node, nodes, allVisited);
    tree.children.push(treeNode);
    currentId = node.next;
  }

  // Find orphan nodes (nodes not in the connected chain)
  const allNodeIds = new Set(Object.keys(nodes).filter((id) => id !== "start"));
  const potentialOrphanIds = new Set(
    [...allNodeIds].filter((id) => !allVisited.has(id)),
  );

  // Collect reporter IDs from ALL potential orphan nodes
  for (const nodeId of potentialOrphanIds) {
    const node = nodes[nodeId] || {};
    collectReporterIds(node.inputs || {}, nodes, allVisited);
  }

  // Recalculate orphan_ids with reporter references accounted for
  const orphanIds = new Set(
    [...allNodeIds].filter((id) => !allVisited.has(id)),
  );

  // Build orphan chains: find chain heads (orphans not pointed to by other orphans)
  const orphanNextTargets = new Set<string>();
  for (const nodeId of orphanIds) {
    const node = nodes[nodeId] || {};
    const nextId = node.next;
    if (nextId && orphanIds.has(nextId)) {
      orphanNextTargets.add(nextId);
    }
  }

  // Chain heads are orphans that no other orphan points to
  const orphanHeads = [...orphanIds]
    .filter((oid) => !orphanNextTargets.has(oid))
    .sort();

  // Process each orphan chain starting from its head
  const processedOrphans = new Set<string>();
  for (const headId of orphanHeads) {
    let currentOrphanId: string | undefined = headId;
    while (currentOrphanId && !processedOrphans.has(currentOrphanId)) {
      const node: NodeDefinition | undefined = nodes[currentOrphanId];
      if (!node) break;

      // Skip reporter nodes (they're embedded in other nodes)
      if (node.isReporter) {
        processedOrphans.add(currentOrphanId);
        currentOrphanId =
          node.next && orphanIds.has(node.next) ? node.next : undefined;
        continue;
      }

      processedOrphans.add(currentOrphanId);
      const orphanTree = nodeToTree(currentOrphanId, node, nodes, allVisited);

      // Include next pointer if it points to another orphan (for chain visualization)
      const nextId = node.next;
      if (nextId && orphanIds.has(nextId)) {
        orphanTree.next = nextId;
      }

      tree.orphans!.push(orphanTree);
      currentOrphanId = nextId && orphanIds.has(nextId) ? nextId : undefined;
    }
  }

  return tree;
}

function nodeToTree(
  nodeId: string,
  node: NodeDefinition,
  allNodes: Record<string, NodeDefinition>,
  allVisited: Set<string>,
): TreeNode {
  const opcode = node.opcode || "";
  const inputs = node.inputs || {};
  const isReporter = node.isReporter || false;

  // Track reporter nodes referenced in inputs
  collectReporterIds(inputs, allNodes, allVisited);

  const treeNode: TreeNode = {
    id: nodeId,
    type: getNodeType(opcode),
    opcode,
    isReporter,
    inputs: formatInputs(inputs, allNodes, opcode),
    children: [],
  };

  // Get construct from grammar for grammar-driven extraction
  const construct = getConstruct(opcode);

  if (construct && construct.branches.length > 0) {
    // Extract config for loops and other constructs with special inputs
    treeNode.config = extractConfig(opcode, inputs, allNodes, construct);

    // Grammar-driven branch extraction
    treeNode.children = extractBranches(inputs, allNodes, allVisited, construct);
  }

  return treeNode;
}

/**
 * Extract config values from construct inputs (for loops, etc.)
 */
function extractConfig(
  opcode: string,
  inputs: Record<string, unknown>,
  allNodes: Record<string, NodeDefinition>,
  construct: Construct,
): LoopConfig {
  const config: LoopConfig = {};

  // Extract values for non-branch inputs
  for (const inputDef of construct.inputs) {
    const value = getInputValue(inputs, inputDef.name);
    if (value !== undefined) {
      const key = inputDef.name.toLowerCase();

      if (inputDef.type === "variable_name") {
        // Variable name inputs are stored as raw string
        config[key] = getRawValue(value) as string;
      } else if (inputDef.type === "expression") {
        // Expression inputs need formatting
        if (key === "condition" || key === "iterable" || key === "resource") {
          config[key] = formatValue(value, allNodes);
        } else {
          // Numeric values (start, end, step, timeout)
          config[key] = getRawValue(value) as number;
        }
      }
    }
  }

  // Legacy support: ensure var is set for loops
  if (["control_for", "control_foreach", "control_spawn", "control_async_foreach", "control_with"].includes(opcode)) {
    if (!config.var) {
      config.var = getRawValue(inputs["VAR"] ?? inputs["var"] ?? "i") as string;
    }
  }

  return config;
}

/**
 * Grammar-driven branch extraction.
 * Handles both static and dynamic branches based on construct definition.
 * Includes empty branches (with branch: null) so slots render in UI.
 */
function extractBranches(
  inputs: Record<string, unknown>,
  allNodes: Record<string, NodeDefinition>,
  allVisited: Set<string>,
  construct: Construct,
): BranchNode[] {
  const branches: BranchNode[] = [];

  if (construct.dynamic_branches) {
    // Handle dynamic branches (try/fork patterns)
    return extractDynamicBranches(inputs, allNodes, allVisited, construct);
  }

  // Standard branch extraction from grammar
  for (const branchDef of construct.branches) {
    const branchInput = getInputValue(inputs, branchDef.name) as InputValue | undefined;
    const branchId = typeof branchInput === "object" && branchInput !== null
      ? branchInput.branch
      : undefined;

    if (branchId) {
      branches.push(buildBranch(branchDef.name, branchId, allNodes, allVisited));
    } else if (branchInput !== undefined) {
      // Empty branch slot (branch: null) - include as empty branch
      branches.push({ type: "branch", name: branchDef.name, children: [] });
    }
  }

  return branches;
}

/**
 * Extract dynamic branches for constructs like try/fork.
 * Uses grammar to identify patterns and extract all matching branches.
 * Includes empty branches (with branch: null) so slots render in UI.
 */
function extractDynamicBranches(
  inputs: Record<string, unknown>,
  allNodes: Record<string, NodeDefinition>,
  allVisited: Set<string>,
  construct: Construct,
): BranchNode[] {
  const branches: BranchNode[] = [];

  // Identify static vs dynamic branch patterns from grammar
  const staticBranchNames: string[] = [];
  const dynamicPrefixes: string[] = [];

  for (const branchDef of construct.branches) {
    const match = branchDef.name.match(/^([A-Z]+)(\d+)$/);
    if (match) {
      const prefix = match[1];
      if (!dynamicPrefixes.includes(prefix)) {
        dynamicPrefixes.push(prefix);
      }
    } else {
      staticBranchNames.push(branchDef.name);
    }
  }

  // Extract static branches first (e.g., TRY, FINALLY)
  for (const branchName of staticBranchNames) {
    const branchInput = getInputValue(inputs, branchName) as InputValue | undefined;
    const branchId = typeof branchInput === "object" && branchInput !== null
      ? branchInput.branch
      : undefined;

    if (branchId) {
      branches.push(buildBranch(branchName, branchId, allNodes, allVisited));
    } else if (branchInput !== undefined) {
      // Empty branch slot (branch: null) - include as empty branch
      branches.push({ type: "branch", name: branchName, children: [] });
    }
  }

  // Extract dynamic branches by pattern
  for (const prefix of dynamicPrefixes) {
    // Check for list format (HANDLERS, BRANCHES)
    const listKey = prefix === "CATCH" ? "HANDLERS" : prefix === "BRANCH" ? "BRANCHES" : null;
    const listInput = listKey ? (inputs[listKey] ?? inputs[listKey.toLowerCase()]) : null;

    if (Array.isArray(listInput) && listInput.length > 0) {
      // List format
      listInput.forEach((item, i) => {
        if (typeof item === "object" && item !== null) {
          const handler = item as CatchHandler;
          const branchId = handler.branch ?? (handler.body as InputValue | undefined)?.branch;
          const branchName = `${prefix}${i + 1}`;

          if (branchId) {
            const branch = buildBranch(branchName, branchId, allNodes, allVisited);
            if (prefix === "CATCH") {
              branch.exception_type = handler.exception_type || "Exception";
              branch.var_name = handler.var_name ?? handler.var;
            }
            branches.push(branch);
          } else {
            // Empty branch slot - include so UI can show the slot
            const emptyBranch: BranchNode = { type: "branch", name: branchName, children: [] };
            if (prefix === "CATCH") {
              emptyBranch.exception_type = handler.exception_type || "Exception";
              emptyBranch.var_name = handler.var_name ?? handler.var;
            }
            branches.push(emptyBranch);
          }
        }
      });
    } else {
      // Individual keys format (CATCH1, CATCH2, BRANCH1, BRANCH2, etc.)
      let i = 1;
      while (true) {
        const key = `${prefix}${i}`;
        const branchInput = inputs[key];
        if (branchInput === undefined) break;

        if (typeof branchInput === "object" && branchInput !== null) {
          const handler = branchInput as CatchHandler;
          // Try different branch reference formats
          const branchId = handler.branch
            ?? (handler.body as InputValue | undefined)?.branch
            ?? (branchInput as InputValue).branch;

          if (branchId) {
            const branch = buildBranch(key, branchId, allNodes, allVisited);
            if (prefix === "CATCH") {
              branch.exception_type = handler.exception_type || "Exception";
              branch.var_name = handler.var_name ?? handler.var;
            }
            branches.push(branch);
          } else {
            // Empty branch slot - include so UI can show the slot
            const emptyBranch: BranchNode = { type: "branch", name: key, children: [] };
            if (prefix === "CATCH") {
              emptyBranch.exception_type = handler.exception_type || "Exception";
              emptyBranch.var_name = handler.var_name ?? handler.var;
            }
            branches.push(emptyBranch);
          }
        }
        i++;
      }
    }
  }

  return branches;
}

/**
 * Get input value with case-insensitive key lookup.
 */
function getInputValue(inputs: Record<string, unknown>, key: string): unknown {
  return inputs[key] ?? inputs[key.toLowerCase()];
}

function collectReporterIds(
  inputs: Record<string, unknown>,
  allNodes: Record<string, NodeDefinition>,
  allVisited: Set<string>,
): void {
  for (const value of Object.values(inputs)) {
    if (typeof value === "object" && value !== null) {
      const inputValue = value as InputValue;
      if (inputValue.node) {
        const nodeId = inputValue.node;
        if (!allVisited.has(nodeId)) {
          allVisited.add(nodeId);
          const node = allNodes[nodeId] || {};
          // Recursively collect from reporter's inputs
          const nodeInputs = node.inputs || {};
          collectReporterIds(nodeInputs, allNodes, allVisited);
        }
      }
    }
  }
}

function getNodeType(opcode: string): NodeType {
  const grammar = getGrammar();
  for (const category of grammar.categories) {
    if (opcode.startsWith(category.prefix)) {
      // Map category id to NodeType
      switch (category.id) {
        case "control":
        case "async":
          return "control_flow";
        case "data":
          return "data";
        case "io":
          return "io";
        case "operator":
        case "math":
        case "string":
        case "list":
        case "dict":
          return "operator";
        case "workflow":
          return "workflow_op";
        default:
          return "opcode";
      }
    }
  }
  return "opcode";
}

/**
 * Format inputs for display, filtering out branch inputs based on grammar.
 */
function formatInputs(
  inputs: Record<string, unknown>,
  allNodes: Record<string, NodeDefinition>,
  opcode: string,
): Record<string, FormattedValue> {
  const formatted: Record<string, FormattedValue> = {};
  const construct = getConstruct(opcode);

  // Build set of branch input names to filter out
  const branchInputNames = new Set<string>();
  if (construct) {
    for (const branch of construct.branches) {
      branchInputNames.add(branch.name);
      branchInputNames.add(branch.name.toLowerCase());
    }
  }

  // Always filter these special keys
  const specialKeys = new Set([
    "BODY", "body", "THEN", "then", "ELSE", "else",
    "TRY", "FINALLY", "HANDLERS", "handlers", "BRANCHES", "branches"
  ]);

  for (const [key, value] of Object.entries(inputs)) {
    // Skip branch inputs
    if (branchInputNames.has(key) || specialKeys.has(key)) {
      continue;
    }
    // Skip dynamic branch patterns
    if (/^(CATCH|BRANCH)\d+$/.test(key)) {
      continue;
    }

    formatted[key] = formatValue(value, allNodes);
  }

  return formatted;
}

function formatValue(
  value: unknown,
  allNodes: Record<string, NodeDefinition>,
  _depth: number = 0,
): FormattedValue {
  if (typeof value === "object" && value !== null) {
    const inputValue = value as Record<string, unknown>;

    if ("literal" in inputValue) {
      return { type: "literal", value: inputValue.literal };
    }
    if ("variable" in inputValue) {
      return { type: "variable", name: inputValue.variable as string };
    }
    if ("node" in inputValue) {
      const nodeId = inputValue.node as string;
      const node = allNodes[nodeId] || {};
      const opcode = node.opcode || "";
      const nodeInputs = node.inputs || {};
      return {
        type: "reporter",
        id: nodeId,
        opcode,
        inputs: Object.fromEntries(
          Object.entries(nodeInputs)
            .filter(
              ([k]) =>
                !k.startsWith("CATCH") && !["BODY", "THEN", "ELSE"].includes(k),
            )
            .map(([k, v]) => [k, formatValue(v, allNodes, _depth + 1)]),
        ),
      };
    }
    if ("branch" in inputValue) {
      return { type: "branch", target: inputValue.branch as string };
    }
    if ("workflow_call" in inputValue) {
      return {
        type: "workflow_call",
        name: inputValue.workflow_call as string,
      };
    }
    return { type: "dict", value };
  }

  return { type: "literal", value };
}

function getRawValue(value: unknown): unknown {
  if (
    typeof value === "object" &&
    value !== null &&
    "literal" in (value as Record<string, unknown>)
  ) {
    return (value as { literal: unknown }).literal;
  }
  return value;
}

function buildBranch(
  name: string,
  startId: string,
  allNodes: Record<string, NodeDefinition>,
  allVisited: Set<string>,
): BranchNode {
  const branch: BranchNode = {
    type: "branch",
    name,
    children: [],
  };

  let currentId: string | undefined = startId;
  while (currentId && !allVisited.has(currentId)) {
    allVisited.add(currentId);
    const node: NodeDefinition | undefined = allNodes[currentId];
    if (!node) break;
    branch.children.push(nodeToTree(currentId, node, allNodes, allVisited));
    currentId = node.next;
  }

  return branch;
}
